### PR TITLE
add support for custom template file default suffix

### PR DESCRIPTION
**Pull request by pine3ree**
PR for: https://github.com/zendframework/zend-expressive/issues/641

I wonder if we should use the same configuration key in all compatible implementations (platesrenderer already support custom suffix with the `extension` config key)

### CODE REVIEW

File: test/TestAsset/zendview-custom-suffix.pht [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<h1>This is a template file for ZendView with <em>php</em> suffix</h1>
```
Comment (weierophinney) [inline test/TestAsset/zendview-custom-suffix.pht]: Should this read "php" or "pht"? the file itself has a "pht" suffix...

**Review by weierophinney — CHANGES_REQUESTED** [review]

File: test/TestAsset/zendview-custom-suffix.pht (hunk 1 above)
Comment (pine3ree) [reply test/TestAsset/zendview-custom-suffix.pht]: yes, right...sorry... it was a ".php" extension originally...
https://github.com/zendframework/zend-expressive-zendviewrenderer/pull/64/commits/6e20aff1a34c1b4d8e5397e79adb2e85ba8a1317

**Comment by Xerkus** [comment]
I do not like much that configuration for the resolver needs to be added to template renderer like that. It feels out of place here tbh. I have no other suggestions at this point however.

**Comment by pine3ree** [comment]
Hello @Xerkus 
> I do not like much that configuration for the resolver needs to be added to template renderer like that. It feels out of place here tbh. I have no other suggestions at this point however.

I added the config entry for completeness, but it can be omitted because the default template resolver's suffix is 'phtml'. If a config entry exists, that will be injected into the resolver. Maybe we could comment out that entry with a clarifying note.

kind regards

**Comment by Xerkus** [comment]
@pine3ree let me clarify:
It is not about your change specifically but why you had to do it like that. A code purism issue if you like, in a broader scope than that of your PR.

The fact that you had to modify template renderer to add functionality, the configuration of dependency,  that is not related to template renderer responsibilities is a violation of Separation of Concern principle.

Dependency is internal detail as long as you do not need external input for it, which makes it conditionally acceptable. Unit testing usually is a first place where the problem is exposed, when you need to provide alternative implementation for testing purposes.
Dependency Injection is a technique that provides separation of responsibilities by moving setup and configuration of dependency to external location. Dependency Injection Container, or simply Container, is usually that location.

The proper solution here is to remove dependency setup from template renderer and do that in a factory. That would, however, constitute a Backward Compatibility break so it would need a major version release to happen. 

Now why it was done like that in the first place? Convenience of use was overriding factor when this code was initially implemented, it needed to just work and we had no convenient way to provide and configure container factories. It was before zend-config-aggregator component to consume ConfigProviders, zend-component-installer component to auto-inject ConfigProviders and config adapters for various container implementations.

So my dislike of the situation is just that: a dislike. I considered other possible approaches and your offers the least impact on the future compatibility with the next major release. That makes it the best we could do right now.

File: test/ZendViewRendererFactoryTest.php [hunk 2]
```diff
@@ -247,6 +248,34 @@ public function testConfiguresTemplateMap()
...
+
+        $factory = new ZendViewRendererFactory();
+        $view = $factory($this->container->reveal());
+
+        $r = new ReflectionProperty($view, 'resolver');
+        $r->setAccessible(true);
+        $resolver  = $r->getValue($view);
+
+        // @codingStandardsIgnoreStart
+        // phpcs:disable
```
Comment (Xerkus) [inline test/ZendViewRendererFactoryTest.php]: please remove CS ignore, make assert call multiline

File: test/ZendViewRendererTest.php [hunk 3]
```diff
@@ -607,6 +607,18 @@ public function testRenderChildWithDefaultParameter()
         static::assertEquals($content, $result);
     }
 
+    public function testCanRenderWithCustomSuffix()
```
Comment (Xerkus) [inline test/ZendViewRendererTest.php]: test name is a bit ambiguous as custom suffix could be specified as part of template name, please rename to
```suggestion
    public function testCanRenderWithCustomDefaultSuffix()
```

**Review by Xerkus — CHANGES_REQUESTED** [review]

File: test/ZendViewRendererFactoryTest.php (hunk 2 above)
Comment (Xerkus) [reply test/ZendViewRendererFactoryTest.php]: @pine3ree  You got the wrong one. This one is not fixed yet.

Comment (pine3ree) [reply test/ZendViewRendererFactoryTest.php]: sorry...late hour here and busy day....do I revert back the other one? ....I was wondering why ruining the "series" ....now I know why... :-)

Comment (Xerkus) [reply test/ZendViewRendererFactoryTest.php]: Nah, other one is good. keep it

Comment (pine3ree) [reply test/ZendViewRendererFactoryTest.php]: oh...gosh...sorry... "keep it"? Keep my commit (https://github.com/zendframework/zend-expressive-zendviewrenderer/pull/64/commits/c27c053c6e11957c9d742581f513da50ed73b339), or keep the codingStandardsIgnore as it was before? Please just write me "REVERT!" if you mean this....

Comment (Xerkus) [reply test/ZendViewRendererFactoryTest.php]: keep the commit.
I'll fix this one and merge.

**Review by Xerkus — APPROVED** [review]

File: test/ZendViewRendererFactoryTest.php (hunk 2 above)
Comment (pine3ree) [reply test/ZendViewRendererFactoryTest.php]: ah ok, I had them ready....I was 1 second from pushing to my repo...thanks

**Comment by pine3ree** [comment]
Hello @Xerkus,
I agree with everything you wrote (about separation of concerns in code), except one thing, if I understood correctly.

> The fact that you had to modify template renderer to add functionality, the configuration of dependency, that is not related to template renderer responsibilities is a violation of Separation of Concern principle.

If you are referring to zend-expressive ZendViewRenderer, in my opinion it is not just a template "renderer" but an implementation of a very simple interface acting as a bridge to a whole package. zend-view in this case....well... parts of it.

zend-expressive template renderer interface has 4 methods: 

- 1 for rendering the template (this implies being able to resolving it by name w/o an expressive resolver interface)
- 2 related to template paths
- 1 for adding default params into a template  

so the interface itself tells us that ii is not just a "pure renderer" interface

Internally zend-expressive-zendviewrenderer creates the zend-view (php-)renderer, the resolver, the view model and allows to add search paths...and more...so adding a configuration parameter in the constructor only exploits what it is already doing....i.e. ... quite a lot....no much separation of concerns...the separation of concerns is implemented in the package it refers to! 
The alternative would be to add more interfaces (renderer, resolver, model) for different concerns to the template part of expressive, but that would make things maybe incompatible to other renderer packages it now supports. 

kind regards,
maks

**Comment by Xerkus** [comment]
@pine3ree nothing in this repository provides interoperability contract. So, anything that goes beyond what is established by `Zend\Expressive\Template\TemplateRendererInterface` is this adapter specific, including resolver and its configurations. As such there is no need to keep everything wrapped in template renderer. 

I am going to provide a patch soon-ish that would resolve my concerns for the next major. You can ping me in zendframework slack and we can discuss this in detail if you like.

**Comment by pine3ree** [comment]
@Xerkus,
sure, thanks!
kind regards

**Pull request merged**